plugins/retry.js

```diff
@@ -32,48 +30,63 @@ module.exports = function(options) {
     var maxAttempts = options.retryAttempts || 3;
     var firstTimeout = options.retryTimeout || 500; // ms
     var timeout = 0; // ms
-    var statusCode = null;
-
-    if (typeof callback !== 'function') {
-      callback = nullcallback;
-    }
+    var retry;
 
     // create a pass-through stream in case the caller wishes
     // to pipe data using Node.js streams
     var s = new stream.PassThrough();
 
-    // do the first function until the second function returns true
-    async.doUntil(function(done) {
-      statusCode = 500;
+    // add error listener
+    s.on('error', function(err) {
+      debug(err);
+    });
+
+    // do the first function until the second function returns false
+    async.doWhilst(function(done) {
       attempts++;
+      retry = false;
+
       if (attempts >= 1) {
         debug('attempt', attempts, 'timeout', timeout);
       }
+
       setTimeout(function() {
-        request(req, function(e, h, b) {
-          done(null, [e, h, b]);
-        }).on('response', function(r) {
-          statusCode = (r && r.statusCode) || 500;
-        }).on('data', function(chunk) {
-          if (statusCode !== 429) {
-            s.write(chunk);
-          }
-        });
+        var thisRequest = request(req, callback);
+        thisRequest
+          .on('error', function(err) {
+            s.emit('error', err);
+            s.end();
+            done();
+          })
+          .on('response', function(r) {
+            if (r.statusCode === 429 && attempts < maxAttempts) {
```

Comment:
Some of our other libraries have an increasing timeout between 429 failures, do you think this is something we should do to be consistent with them?